### PR TITLE
[Clement Tee] ip

### PR DESCRIPTION
**Pull request by clement0010**
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/duke/storage/Storage.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            throw new DukeException("duke.storage.Storage file should end with '.txt'");
+        }
+    }
+
+    public ArrayList<Task> loadALlTasks() throws DukeException  {
+        ArrayList<Task> taskList = new ArrayList<Task>();
+        if (isValidPath(path)) {
+            return taskList;
+        }
+        try(BufferedReader reader = new BufferedReader(new FileReader(path.toString()))) {
```
Comment (Gernene) [inline src/main/java/duke/storage/Storage.java]: I think you're missing a tiny space here.

File: src/main/java/duke/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.Arrays;
+
+public class Parser {
+    private static final String DEFAULT_DATE_FORMAT = "d/MM/yyyy HHmm";
+
+    public static Command parse(String userInput) throws DukeException {
```
Comment (Gernene) [inline src/main/java/duke/parser/Parser.java:34]: This is a total nitpick, but don't forget to add your header comments! xD

File: src/main/java/duke/command/AddTodoCommand.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package duke.command;
+
+import duke.common.Messages;
+import duke.storage.Storage;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.Todo;
+import duke.ui.TextUi;
+
+public class AddTodoCommand extends Command {
```
Comment (kerburenthusiasm) [inline src/main/java/duke/command/AddTodoCommand.java:14]: Perhaps follow a consistent style in naming classes extending `Command` class?
```suggestion
public class TodoCommand extends Command {
```

**Review by Gernene — COMMENTED** [review]
Hi! I couldn't find many coding standard items to fix. I left a few nitpick markers here and there, but otherwise looks good. Nice work!

File: src/main/java/duke/command/Command.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import duke.DukeException;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.TextUi;
+
+public abstract class Command {
+    private final boolean isExit;
+
+    public Command (boolean isExit) {
```
Comment (kerburenthusiasm) [inline src/main/java/duke/command/Command.java]: Minor change needed here.

```suggestion
    public Command(boolean isExit) {
```

File: src/main/java/duke/task/Task.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            return "E";
+        } else if (Deadline.class.equals(this.getClass())) {
+            return "D";
+        } else {
+            return " ";
+        }
+
+    }
+
+    public boolean contains(String keywords) {
```
Comment (kerburenthusiasm) [inline src/main/java/duke/task/Task.java]: Perhaps a function name which sounds more like a boolean?
```suggestion
    public boolean hasKeywords(String keywords) {
```

File: src/test/java/DukeTest.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import duke.parser.Parser;
+import duke.task.Task;
+import duke.task.Todo;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DukeTest {
+    @Test
+    public void testTaskSizeParser() {
```
Comment (kerburenthusiasm) [inline src/test/java/DukeTest.java]: Perhaps a more descriptive name for the test function?

```suggestion
    public void testTaskSizeParser_emptyString() {
```

File: src/main/java/duke/task/Task.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    public String getClassSymbol() {
+        if (Todo.class.equals(this.getClass())) {
+            return "T";
+        } else if (Event.class.equals(this.getClass())) {
+            return "E";
+        } else if (Deadline.class.equals(this.getClass())) {
+            return "D";
+        } else {
+            return " ";
+        }
```
Comment (kerburenthusiasm) [inline src/main/java/duke/task/Task.java]: Perhaps throw an exception for the default case?

```suggestion
        } else {
            throw new DukeException("Unknown Task");
        }
```

File: src/main/java/duke/task/TaskList.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException(e.getMessage());
+        }
+    }
+
+    public void addTask(Task task) {
+        tasks.add(task);
+    }
+
+    public TaskList filterTask(String keywords) {
```
Comment (kerburenthusiasm) [inline src/main/java/duke/task/TaskList.java]: Perhaps a function name which sounds more like a plural as the function is filtering among many tasks?

```suggestion
    public TaskList filterTasks(String keywords) {
```

File: src/main/java/duke/storage/Storage.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public Storage(String fileName) throws DukeException {
+        path = Paths.get(DEFAULT_STORAGE_DIRECTORY + fileName);
+        if (isValidPath(path)) {
+            throw new DukeException("duke.storage.Storage file should end with '.txt'");
+        }
+    }
+
+    public ArrayList<Task> loadALlTasks() throws DukeException  {
```
Comment (kerburenthusiasm) [inline src/main/java/duke/storage/Storage.java]: Minor mistake found in function name.

```suggestion
    public ArrayList<Task> loadAllTasks() throws DukeException  {
```

File: src/main/java/duke/command/FindTaskCommand.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        this.keywords = keywords;
+    }
+
+    @Override
+    public void execute(TaskList tasks, TextUi ui, Storage storage) {
+        try {
+            TaskList filteredTasks = tasks.filterTask(keywords);
+            for (int i = 0; i < filteredTasks.getSize(); i++) {
+                ui.showToUser((i + 1) + ". " + filteredTasks.getTask(i));
+            }
```
Comment (kerburenthusiasm) [inline src/main/java/duke/command/FindTaskCommand.java]: Perhaps encapsulate this part of the code into a function to reduce code duplication?

```suggestion
            printFilteredTasks();
```

**Review by kerburenthusiasm — COMMENTED** [review]
Overall, the code is readable and strictly follows Java coding standard. However, some of the files requires header comments for the classes and functions. I noted a few minor coding standard violations.

File: src/main/java/duke/common/Messages.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    public static final String MESSAGE_UNMARKED = "OK, I've marked this task as not done yet.";
+    public static final String MESSAGE_DELETE = "Noted. I've removed this task.";
+    public static final String MESSAGE_FIND = "Here are the matching tasks in your list";
+    public static final String MESSAGE_ADD_TODO = "Got it. I've added this todo.";
+    public static final String MESSAGE_ADD_DEADLINE = "Got it. I've added this deadline.";
+    public static final String MESSAGE_ADD_EVENT = "Got it. I've added this event.";
+    public static final String MESSAGE_ERROR_INVALID_COMMAND = "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
+    public static final String MESSAGE_ERROR_EMPTY_TITLE = "☹ OOPS!!! The description of a todo cannot be empty.";
+    public static final String MESSAGE_ERROR_INVALID_DATETIME_FORMAT= "Invalid date time format.";
+}
```
Comment (arcornior) [inline src/main/java/duke/common/Messages.java]: Good job with the constant naming convention!

File: src/main/java/duke/parser/Parser.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                userInput = String.join(" ",
+                        Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
+                return prepareAddEventCommand(userInput);
+            case "deadline":
+                userInput = String.join(" ",
+                        Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
+                return prepareAddDeadlineCommand(userInput);
+            default:
+                throw new DukeException(Messages.MESSAGE_ERROR_INVALID_COMMAND);
+        }
```
Comment (arcornior) [inline src/main/java/duke/parser/Parser.java]: Shouldn't the case indent be on the same level as the switch statement?
I noticed that this issue occurred a few times in your program. Do consider taking note of this.
```suggestion
        switch (userInputArr[0]) {
        case "mark":
            return new MarkTaskCommand(Integer.parseInt(userInputArr[1]) - 1, true);
        case "unmark":
            return new MarkTaskCommand(Integer.parseInt(userInputArr[1]) - 1, false);
        case "delete":
            return new DeleteCommand(Integer.parseInt(userInputArr[1]) - 1);
        case "find":
            userInput = String.join(" ",
                    Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
            return new FindTaskCommand(userInput);
        case "list":
            return new ListCommand();
        case "bye":
            return new ExitCommand();
        case "todo":
            return prepareAddTodoCommand(userInputArr[1]);
        case "event":
            userInput = String.join(" ",
                    Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
            return prepareAddEventCommand(userInput);
        case "deadline":
            userInput = String.join(" ",
                    Arrays.copyOfRange(userInputArr, 1, userInputArr.length));
            return prepareAddDeadlineCommand(userInput);
        default:
            throw new DukeException(Messages.MESSAGE_ERROR_INVALID_COMMAND);
        }
```

File: src/main/java/duke/command/ListCommand.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+public class ListCommand extends Command {
+    private static final boolean IS_EXIT = false;
+
+    public ListCommand() {
+        super(IS_EXIT);
+    }
+
+    @Override
+    public void execute(TaskList tasks, TextUi ui, Storage storage) {
```
Comment (arcornior) [inline src/main/java/duke/command/ListCommand.java]: Should there be a JavaDocs for this method?

File: src/main/java/duke/ui/TextUi.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
+package duke.ui;
+
+import duke.common.Messages;
+import duke.parser.Parser;
+
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.Scanner;
```
Comment (arcornior) [inline src/main/java/duke/ui/TextUi.java]: Great job separating your import statements!

**Review by arcornior — COMMENTED** [review]
LGTM! Just a few nits to fix.

**Pull request closed without merging**